### DISCUSSION TITLE
Android support

### DISCUSSION THREAD

**Discussion by MyelinsheathXD** [google-deepmind/mujoco#1638] [Ideas]
Hello everybody!
Is there any future plans to support android support for mujoco engine?
I have tried locally, yet stuck at c++ native function “qsortr “array sorting function conversion.
It seems android do not have native c++ library function for “qsortr” function.

**Comment by yuvaltassa** [comment]
Ah this is interesting. There are other reasons to stop using these library sorting functions, so it's good you brought this up.

That said, I'm very surprised this basic function is unavailable. Surely you can find a relevant discussion on stackexchange or similar?

**Comment by yuvaltassa** [comment]
Actually hold on, are you trying to run the compiler or just the engine? I don't think the engine calls this function.

**Comment by MyelinsheathXD** [comment]
I was trying to compile the engine as dynamic library using Android Studio's Cmake . 
So far I have successfully port all  3rd party libraries(qhulstatic_r, ccd,tinyxml2,tinyobjectloader,pngloader) into static android library(.a).
Yet, the qsort funtion  from macro mjQUICKSORT and other bugs were stopping me to build the engine for android.
Thank you clarifiing the this bug. it seems very few engine functions use this function.
I will update   if solution for for this bug solves the main build issue.

**Comment by jaraujo98** [comment]
@MyelinsheathXD @yuvaltassa @Balint-H I was able to compile MuJoCo for Android. From my research, `qsort_r` is a non-standard function so it is up to the standard library implementation to provide it (which the Android developers do not do). I borrowed the implementation from GNU (LGPL 2.1 or later), fixed an extra preprocessor directive, provided the appropriate arguments to CMake, and from there it was smooth sailing to having a Unity app running MuJoCo natively on Quest 2. I'm happy to contribute the fixes and the documentation, but I wanted to hear your thoughts on my approach before doing so.

**Reply by Balint-H** [reply]
That's great! So to double check, this would imply a separate android compatible library file for releases as well, right? 

Did you need to make any changes to the Unity plugin?

**Reply by jaraujo98** [reply]
No changes to the Unity plugin (thankfully...!). Just move the new `so` library file to the plugin folder after import, rename it from `libmujoco.so` to `mujoco.so`, and use the editor to specify that it was for Android. This even allows it to coexist with the Mac `dylib` file (which I also marked as Mac only), so I get to test on my Mac using native binaries, and deploy to Android using native binaries. Hard to beat! :-D

It would require a separate Android library file to be released, but that is easy to do (I have a build script I can share). The only pain point I see is that the Android API, architecture, and ABI are locked at compile time, and I can't tell what are sensible defaults (besides the ones I used for the Quest 2 build).

**Reply by yuvaltassa** [reply]
This sounds great but to clarify, are these just changes to the cmake, or do you propose to add this function to the codebase? We cannot redistribute LGPL code.

**Reply by jaraujo98** [reply]
The function needs to be added to the code base. On the other hand, it should be somewhat straightforward to write a wrapper around the normal `qsort` function (which should be available on all platforms as it is part of the standard) that does away with `qsort_r` and `qsort_s` (unfortunately I don't have the bandwidth to help with this right now). Or, if you have a plan to replace library sorting functions, implementing it should also solve it.

**Reply by yuvaltassa** [reply]
I think the right thing is to write a C wrapper with the same API for `std::sort` since `qsort` is not thread safe (which is why `qsort_r` exists). Would that approach work for you?

**Reply by jaraujo98** [reply]
I think I understand the general idea of your approach, and I agree with it. However, I'm not sure if I would be able to implement it without either re-implementing `qsort_r` from scratch, or using a custom `struct` to pass the elements together with the context to the comparison function so that it becomes compatible with `qsort` and sidesteps the thread safety issue (which I'm afraid I do not fully understand). How were you thinking of going about it?

Either way, the only thing preventing MuJoCo from running natively on Android is an implementation of `quicksortfunc` and `mjQUICKSORT` that does not depend on `qsort_r`, but as I mentioned previously I currently lack the time to help with this...

**Reply by joanllobera** [reply]
Hi @jaraujo98 we are looking into doing this and would like to maintain an opensource version of it, if possible. 

we would like to test it with 2 or 3 different android-based headset models. Would you be so kind to detail the changes in the CMAKE as well as the preprocessor directive that you modified? 

we would try it on the Quest 2 that oyu used, the Quest Pro and also the new Vive Focus Vision, provide feedback, and release the build open source for anyone who also wants to do it.

**Comment by thowell** [comment]
`mjQUICKSORT` now wraps `std::sort` when compiling MuJoCo as C++ code. Please see [engine_sort.h](https://github.com/google-deepmind/mujoco/blob/35a834f0ba1330f46502b5e37bb0255d568b18f5/src/engine/engine_sort.h#L51).

**Comment by kbayes** [comment]
There's now a native sorting function that replaces qsort_r and std::sort. This should give a nice performance improvement as well.

**Comment by fayezsalka** [comment]
So any help on how to actually build mujoco for Android? What changes we need to have in the cmake to build for Android? Esp now that qsort_r issue is fixed in latest engine?

**Reply by saran-t** [reply]
Is there a specific problem that you've run into? We haven't tried to build it ourselves but I don't see any blocker necessarily.

**Reply by jaraujo98** [reply]
For reference, these are the commands I was using (clone the MuJoCo repo and run it from the root folder). I haven't tested it with the latest version of MuJoCo, but if the `qsort_r` issue is fixed then it should compile. You'll need to set `ANDROID_NDK_HOME`, which Unity installs for you.

```
mkdir build && cd build
cmake -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK_HOME/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_PLATFORM=android-31 -DMUJOCO_BUILD_SIMULATE=OFF -DMUJOCO_BUILD_EXAMPLES=OFF -DMUJOCO_BUILD_TESTS=OFF -DMUJOCO_TEST_PYTHON_UTIL=OFF ..
cmake --build .
```

**Comment by joanllobera** [comment]
Binaries of mujoco 3.3.0 for android for unity projects can be found in this repository: https://github.com/joanllobera/mujoco-bin/
It is very easy ti integrate in a new project. Check the README there.

The android build was actually done by  Yves Schmid-Dornbierer, the instructions to do so can be found in this mujoco fork: https://github.com/yvesantzero/mujoco/blob/main/build_mujoco_android.pdf 


An example project using this can be found in this branch of modular agents: https://github.com/Balint-H/modular-agents/tree/fix/upgrade2mj330 

I hope this is enough for anyone wanting to have mujoco running on android-based VR or on an android phone.

**Reply by joanllobera** [reply]
As an update, this is now updated to mujoco 3.5.0, with the same procedure. Find it here: https://github.com/joanllobera/mujoco-bin/releases/tag/3.5.0

**Reply by lab-key** [reply]
I guess yours is even more cross platform? Unity builds for iOS? 

What I did was tried to integrate it directly into upstream. So you can basically

```
git clone mujoco
./gradlew installRelease
```

And have the Simulate app running on your phone basically

I tried doing it without any external deps but writing EGL figuring rendering / everything out and getting it to work didn't really pan out the best. And the logcat / logging is not exactly providing ideal outputs. I gave up fiddling and just made the rendering in Java. 

So just building libmujoco and using it I wish I saw this earlier. Probably would've given me a solid base, it's probably cleaner than what I have going on...

**Comment by lab-key** [comment]
I had issues with linking rendering and I am doing it on Linux, but I've managed to get it to somewhat work. It's pretty sloppy so as soon as I clean up and standardise things I'll probably open source large parts of it. The thing is adding it into the main repo will be a bit icky as it will require java and I am yet to figure out qhull ' The Official MuJoCo Way ' as I see emscripten build is running into issues too but I am yet to analyze and try out other things.

I'm still learning about a lot so I don't want to act like some know-it-all but the rendering is a nightmare...

[![Video](https://img.youtube.com/vi/vfx9lhYEcV4/0.jpg)](https://www.youtube.com/embed/vfx9lhYEcV4)

[![Video](https://img.youtube.com/vi/72JILVU8FqI/0.jpg)](https://www.youtube.com/shorts/72JILVU8FqI)
